pkg/streamer/file.go

```diff
@@ -129,100 +127,209 @@ func CollectBinlogFilesCmp(dir, baseFile string, cmp FileCmp) ([]string, error)
+func readDir(dirpath string) ([]string, error) {
+	dir, err := os.Open(dirpath)
 	if err != nil {
-		return pos, errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
+	defer dir.Close()
 
-	sepIdx := strings.Index(parsed.baseName, posUUIDSuffixSeparator)
-	if sepIdx > 0 && sepIdx+len(posUUIDSuffixSeparator) < len(parsed.baseName) {
-		return mysql.Position{
-			Name: fmt.Sprintf("%s%s%s", parsed.baseName[:sepIdx], baseSeqSeparator, parsed.seq),
-			Pos:  pos.Pos,
-		}, nil
+	names, err := dir.Readdirnames(-1)
+	if err != nil {
+		return nil, errors.Annotatef(err, "dir %s", dirpath)
 	}
 
-	return pos, nil
-}
+	sort.Strings(names)
 
-type binlogFile struct {
-	baseName string
-	seq      string
+	return names, nil
 }
 
-func (f *binlogFile) BiggerThan(other *binlogFile) bool {
-	return f.baseName == other.baseName && f.seq > other.seq
+// fileSizeUpdated checks whether the file's size has updated
+// return
+//   0: not updated
+//   1: update to larger
+//  -1: update to smaller, only happens in special case, for example we change
+//      relay.meta manually and start task before relay log catches up.
+func fileSizeUpdated(path string, latestSize int64) (int, error) {
+	fi, err := os.Stat(path)
+	if err != nil {
+		return 0, errors.Annotatef(err, "get stat for relay log %s", path)
+	}
+	currSize := fi.Size()
+	if currSize == latestSize {
+		return 0, nil
+	} else if currSize > latestSize {
+		log.Debugf("[streamer] size of relay log file %s has changed from %d to %d", path, latestSize, currSize)
+		return 1, nil
+	} else {
+		log.Errorf("[streamer] size of relay log file %s has changed from %d to %d", path, latestSize, currSize)
+		return -1, nil
+	}
 }
 
-func (f *binlogFile) BiggerOrEqualThan(other *binlogFile) bool {
-	return f.baseName == other.baseName && f.seq >= other.seq
-}
+// relaySubDirUpdated checks whether the relay sub directory updated
+// including file changed, created, removed, etc.
+func relaySubDirUpdated(ctx context.Context, watcherInterval time.Duration, dir string,
+	latestFilePath, latestFile string, latestFileSize int64) (string, error) {
+	// create polling watcher
+	watcher2 := watcher.NewWatcher()
 
-func (f *binlogFile) LessThan(other *binlogFile) bool {
-	return f.baseName == other.baseName && f.seq < other.seq
-}
+	// Add before Start
+	// no need to Remove, it will be closed and release when return
+	err := watcher2.Add(dir)
+	if err != nil {
+		return "", errors.Annotatef(err, "add watch for relay log dir %s", dir)
+	}
 
-// [0-9] in string -> [48,57] in ascii
-func allAreDigits(s string) bool {
-	for _, r := range s {
-		if r >= 48 && r <= 57 {
-			continue
+	err = watcher2.Start(watcherInterval)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	defer watcher2.Close()
+
+	type watchResult struct {
+		updatePath string
+		err        error
+	}
+
+	result := make(chan watchResult, 1) // buffered chan to ensure not block the sender even return in the halfway
```

Comment:
not necessary because when the channel is no longer used, it will be garbage collected.

from another guy also called Ian (https://groups.google.com/d/msg/golang-nuts/pZwdYRGxCIk/qpbHxRRPJdUJ)
> Note that it is only necessary to close a channel if the receiver is
looking for a close.  Closing the channel is a control signal on the
channel indicating that no more data follows.

Reply:
ok